ajour: init at 0.6.3

mweinelt opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

A cross-plattform addon manager for World of Warcraft.

https://github.com/casperstorm/ajour

cc @MetaDark @cawilliamson

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

mweinelt committed:
ajour: init at 0.6.3

mweinelt commented:
It starts up, looks pretty, but I can't select a game directory just yet.

abathur commented:
Off-topic: You may already know this, but I guess it's a recent-ish addition that I only noticed this fall myself, and the peace-of-mind it brings is nice.

There's an option hiding in a dropdown to the right of the Create pull request button--and then hidden away under the reviewers section of the PR info--for marking it as draft:

<img width="357" alt="Screen Shot 2021-01-07 at 7 12 02 PM" src="https://user-images.githubusercontent.com/2548365/103962294-41cfbe80-511c-11eb-810a-4fd0f23abc57.png">

<img width="406" alt="Screen Shot 2021-01-07 at 7 03 49 PM" src="https://user-images.githubusercontent.com/2548365/103962050-8f97f700-511b-11eb-8a6d-20dbd5e776be.png">

mweinelt commented:
Thanks, marked as draft, and your post as off-topic, as you said yourself.

MetaDark commented on pkgs/tools/games/ajour/default.nix:
```diff
@@ -0,0 +1,68 @@
+{ lib
+, fetchFromGitHub
+, rustPlatform
+, autoPatchelfHook
+, cmake
+, pkg-config
+, expat
+, openssl
+, libxkbcommon
+, vulkan-headers
+, vulkan-loader
+, wayland
+, x11
+, xorg
+, python3
```
personal preference nitpick: input order should roughly reflect usage order

MetaDark commented on pkgs/tools/games/ajour/default.nix:
```diff
@@ -0,0 +1,68 @@
+}:
+
+let
+  rpathLibs = [
+    libxkbcommon
+    vulkan-loader
+    wayland
+  ];
```
I noticed that X11 wasn't listed as a load time dependency through ldd, so I switched over to sway and found that these also need to be added to `rpathLibs` for X11 support:
```suggestion
  rpathLibs = [
    libXcursor
    libXi
    libxkbcommon
    libXrandr
    vulkan-loader
    wayland
    x11
  ];
```

mweinelt replied:
I indeed only tested this on wayland at this point. Thanks for coming up with the list for X11.

kira-bruneau reviewed (changes requested):
Thanks @mweinelt! I don't actually play WoW (even though I still have it installed :stuck_out_tongue:), but I can still help review this.

Would the game directory selection normally open a GTK / Qt file dialog? You might need to specify some of the environment variables that are normally set in `wrapGAppsHook` or `wrapQtAppsHook`.

MetaDark commented:
@mweinelt It looks like `native-dialog` requires either zenity or kdialog to be in `PATH`: https://github.com/balthild/native-dialog-rs/blob/master/src/dialog_impl/gnu/mod.rs.

SuperSandro2000 commented on pkgs/tools/games/ajour/default.nix:
```diff
@@ -0,0 +1,68 @@
+    '';
+    homepage = "https://github.com/casperstorm/ajour";
+    changelog = "https://github.com/casperstorm/ajour/blob/master/CHANGELOG.md";
+    licenses = license.mit;
```
```suggestion
    license = license.mit;
```

MetaDark replied:
Nice catch! It also looks like `license.mit` should be changed to `licenses.mit`. It seems like the words were just swapped.

mweinelt commented on pkgs/tools/games/ajour/default.nix:
```diff
@@ -0,0 +1,68 @@
+    openssl
+    xorg.libxcb
+    x11
+  ];
```
Thanks for the clarification.

mweinelt commented:
> Would the game directory selection normally open a GTK / Qt file dialog? You might need to specify some of the environment variables that are normally set in `wrapGAppsHook` or `wrapQtAppsHook`.

Yeah, that was it. I went ahead and put zenity and kdialog in the PATH and it is working now.

MetaDark commented:
Result of `nixpkgs-review pr 108738` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ajour</li>
  </ul>
</details>

piegamesde commented:
When I run the package on Arch Linux, it simply does nothing and exits with code 101 right away :(

The strace line `write(3, "02:25:59.712 [panic][ERROR] thre"..., 453) = 453` suggests that the code aborts during a panic.

MetaDark commented:
@piegamesde Do you know if this panic is caused by a missing runtime library? You can check with `LD_DEBUG=libs ajour`

piegamesde commented:
So, apparently exit code 101 is the panic exit code. So if this is not an abort, why am I not seeing a backtrace?

I don't know how `LD_DEBUG=libs` output looks like on a failure, but I can't see any obviously failing things there.

MetaDark commented:
@piegamesde It looks like ajour uses [log-panics](https://docs.rs/log-panics/2.0.0/log_panics/) to capture panic messages to a log file. Looking through the source, it seems like it should be writing them to `~/.config/ajour/ajour.log`.

piegamesde commented:
@MetaDark Thanks. That explains why the strace command showed the application writing to FD 3 instead of 2 … yeah, there's definitely something of interest in there:

```
02:38:15.342 [ajour][INFO] Ajour 0.6.1 has started.
02:38:15.342 [ajour::gui][DEBUG] config loaded:
Config {
    wow: Wow {
        directory: None,
        flavor: Retail,
    },
    addons: Addons {
        global_release_channel: Stable,
        ignored: {},
        release_channels: {},
        delete_saved_variables: false,
    },
    theme: None,
    column_config: V1 {
        local_version_width: 150,
        remote_version_width: 150,
        status_width: 85,
    },
    window_size: None,
    scale: None,
    backup_directory: None,
    backup_addons: false,
    backup_wtf: false,
    hide_ignored_addons: false,
    self_update_channel: Stable,
    weak_auras_account: {},
    alternating_row_colors: false,
}
02:38:15.342 [ajour::gui][DEBUG] antialiasing: true
02:38:15.358 [panic][ERROR] thread 'main' panicked at 'running Ajour gui: GraphicsAdapterNotFound': src/gui/mod.rs:1053
   0: log_panics::init::{{closure}}
   1: std::panicking::rust_panic_with_hook
   2: rust_begin_unwind
   3: core::panicking::panic_fmt
   4: core::result::unwrap_failed
   5: ajour::gui::run
   6: ajour::main
   7: std::rt::lang_start::{{closure}}
   8: std::rt::lang_start_internal
   9: main
  10: __libc_start_main
  11: _start
```

mweinelt commented:
> 02:38:15.358 [panic][ERROR] thread 'main' panicked at 'running Ajour gui: GraphicsAdapterNotFound': src/gui/mod.rs:1053

Funny, that's basically how I started out, before I started patching the RPATH. Can you test the latest change? It adds all the libraries that I set the RPATH for to buildInputs.

piegamesde commented:
Nope, no change :/ maybe I should open a bug upstream asking for better log diagnostics …

mweinelt commented:
@piegamesde Can you strace the process and check if it is unable to open any libraries?

SuperSandro2000 commented:
Result of `nixpkgs-review pr 108738` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>ajour</li>
  </ul>
</details>

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 108738` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>ajour</li>
</ul>
</details>

MetaDark commented:
Result of `nixpkgs-review pr 108738` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ajour</li>
  </ul>
</details>

mweinelt commented:
Something is really weird.

> anonymous function at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/pkgs/tools/games/ajour/default.nix:1:1 called without required argument 'kdialog', at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/lib/customisation.nix:69:16

```
❯ nix-build -A kdialog
/nix/store/v1gs4i0akskajmin3kvid5pvxa15zkda-kdialog-20.08.3
```

I renamed this from `kdeApplications.kdialog` as of a few days ago, as I couldn't find that anymore, and still can't.

```
❯ nix-build -A kdeApplication.kdialog
error: --- AttrPathNotFound ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- nix-build
attribute 'kdeApplication' in selection path 'kdeApplication.kdialog' not found
```

I've just rebased this onto master and it still builds for me with just `kdialog`. I am clueless.

piegamesde commented:
The kdeApplications problem was introduced by #102168

mweinelt commented:
Migrated to `plasma5Packages.kdialog. Thanks for looking this up.

SuperSandro2000 commented on pkgs/tools/games/ajour/default.nix:
```diff
@@ -0,0 +1,81 @@
+, expat
+, freetype
+, gnome3
+, plasma5Packages
```
Please don't input plasma5Packages. It should be inherit from all-packages.

mweinelt replied:
Can you reference some docs? Why is it fine for gnome3?

SuperSandro2000 replied:
I saw that on a recent comment from ttuegel but I probably won't find it.

SuperSandro2000 commented on pkgs/tools/games/ajour/default.nix:
```diff
@@ -0,0 +1,81 @@
+, python3
+, expat
+, freetype
+, gnome3
```
Same here.
